File: native-ios/RideTrackerNative/TelemetrySourceRouter.swift
import Foundation

struct RoutedTelemetrySample<Value> {
    let metric: String
    let sourceID: String
    let value: Value
    let quality: Double
    let timestamp: TimeInterval
}

struct TelemetrySourcePolicy: Codable, Hashable {
    var metric: String
    var primarySource: String
    var fallbackSources: [String]
    var minimumQuality: Double
    var maxAgeMs: Int
}

struct TelemetrySourceSwitch: Codable, Hashable {
    let timestamp: TimeInterval
    let metric: String
    let from: String?
    let to: String?
    let reason: String
}

@MainActor
final class TelemetrySourceRouter: ObservableObject {
    @Published private(set) var switches: [TelemetrySourceSwitch] = []
    private var latest: [String: (metric: String, value: Any, quality: Double, timestamp: TimeInterval)] = [:]
    private var active: [String: String] = [:]
    var policies: [TelemetrySourcePolicy] = []

    func ingest<Value>(metric: String, sourceID: String, value: Value, quality: Double = 1, timestamp: TimeInterval = ProcessInfo.processInfo.systemUptime) {
        latest[sourceID] = (metric, value, quality, timestamp)
    }

    func resolve<Value>(_ metric: String, at now: TimeInterval = ProcessInfo.processInfo.systemUptime, as type: Value.Type = Value.self) -> RoutedTelemetrySample<Value>? {
        let policy = policies.first { $0.metric == metric }
        let ordered = [policy?.primarySource].compactMap { $0 } + (policy?.fallbackSources ?? [])
        let candidates = ordered.isEmpty ? latest.keys.filter { latest[$0]?.metric == metric } : ordered
        var selected: RoutedTelemetrySample<Value>?
        for source in candidates {
            guard let sample = latest[source], sample.metric == metric, let value = sample.value as? Value else { continue }
            let ageMs = max(0, (now - sample.timestamp) * 1000)
            if sample.quality < (policy?.minimumQuality ?? 0) || ageMs > Double(policy?.maxAgeMs ?? Int.max) { continue }
            selected = RoutedTelemetrySample(metric: metric, sourceID: source, value: value, quality: sample.quality, timestamp: sample.timestamp)
            break
        }
        let next = selected?.sourceID
        if active[metric] != next {
            switches.append(.init(timestamp: now, metric: metric, from: active[metric], to: next, reason: next == nil ? "no-valid-source" : "selected"))
            active[metric] = next
        }
        return selected
    }
}

File: native-android/app/src/main/java/de/ridetracker/sensors/TelemetrySourceRouter.kt
package de.ridetracker.sensors

import android.os.SystemClock

data class TelemetrySourcePolicy(
    val metric:String,
    val primarySource:String,
    val fallbackSources:List<String> = emptyList(),
    val minimumQuality:Double = 0.0,
    val maxAgeMs:Long = Long.MAX_VALUE,
)

data class TelemetrySourceSwitch(
    val timestampMs:Long,
    val metric:String,
    val from:String?,
    val to:String?,
    val reason:String,
)

data class RoutedTelemetrySample<T>(
    val metric:String,
    val sourceId:String,
    val value:T,
    val quality:Double,
    val timestampMs:Long,
)

class TelemetrySourceRouter {
    private data class Stored(val metric:String,val value:Any,val quality:Double,val timestampMs:Long)
    private val latest=mutableMapOf<String,Stored>()
    private val active=mutableMapOf<String,String?>()
    val switches=mutableListOf<TelemetrySourceSwitch>()
    var policies:List<TelemetrySourcePolicy> = emptyList()

    fun reset() {
        latest.clear()
        active.clear()
        switches.clear()
    }

    fun ingest(metric:String,sourceId:String,value:Any,quality:Double=1.0,timestampMs:Long=SystemClock.elapsedRealtime()) {
        latest[sourceId]=Stored(metric,value,quality.coerceIn(0.0,1.0),timestampMs)
    }

    inline fun <reified T> resolve(metric:String,nowMs:Long=SystemClock.elapsedRealtime()):RoutedTelemetrySample<T>? = resolve(metric,nowMs,T::class.java)

    fun <T> resolve(metric:String,nowMs:Long,type:Class<T>):RoutedTelemetrySample<T>? {
        val policy=policies.firstOrNull{it.metric==metric}
        val ordered=(listOfNotNull(policy?.primarySource)+(policy?.fallbackSources?:emptyList())).ifEmpty{latest.filterValues{it.metric==metric}.keys.toList()}
        var selected:RoutedTelemetrySample<T>?=null
        for(source in ordered){
            val sample=latest[source]?:continue
            if(sample.metric!=metric||!type.isInstance(sample.value))continue
            val age=(nowMs-sample.timestampMs).coerceAtLeast(0)
            if(sample.quality<(policy?.minimumQuality?:0.0)||age>(policy?.maxAgeMs?:Long.MAX_VALUE))continue
            selected=RoutedTelemetrySample(metric,source,type.cast(sample.value),sample.quality,sample.timestampMs)
            break
        }
        val next=selected?.sourceId
        if(active[metric]!=next){
            switches+=TelemetrySourceSwitch(nowMs,metric,active[metric],next,if(next==null)"no-valid-source" else "selected")
            active[metric]=next
        }
        return selected
    }
}
